lib/src/io_client.dart

```diff
@@ -23,45 +23,46 @@ class IOClient extends BaseClient {
   IOClient([HttpClient inner]) : _inner = inner ?? new HttpClient();
 
   /// Sends an HTTP request and asynchronously returns the response.
-  Future<StreamedResponse> send(BaseRequest request) async {
-    var stream = request.finalize();
+  Future<Response> send(Request request) async {
+    var stream = await request.read();
 
     try {
       var ioRequest = await _inner.openUrl(request.method, request.url);
+      var context = request.context;
 
       ioRequest
-          ..followRedirects = request.followRedirects
-          ..maxRedirects = request.maxRedirects
+          ..followRedirects = context['io.followRedirects'] ?? true
+          ..maxRedirects = context['io.maxRedirects'] ?? 5
           ..contentLength = request.contentLength == null
               ? -1
               : request.contentLength
-          ..persistentConnection = request.persistentConnection;
+          ..persistentConnection = context['io.persistentConnection'] ?? true;
       request.headers.forEach((name, value) {
         ioRequest.headers.set(name, value);
       });
 
       var response = await stream.pipe(
-          DelegatingStreamConsumer.typed(ioRequest));
+          DelegatingStreamConsumer.typed<List<int>>(ioRequest)
+      ) as HttpClientResponse;
+
       var headers = <String, String>{};
       response.headers.forEach((key, values) {
         headers[key] = values.join(',');
       });
 
-      return new StreamedResponse(
-          DelegatingStream.typed/*<List<int>>*/(response).handleError((error) =>
-              throw new ClientException(error.message, error.uri),
-              test: (error) => error is HttpException),
+      return new Response(
+          _responseUrl(request, response),
           response.statusCode,
-          contentLength: response.contentLength == -1
-              ? null
-              : response.contentLength,
-          request: request,
+          reasonPhrase: response.reasonPhrase,
+          body: DelegatingStream.typed<List<int>>(response).handleError(
+              (error) => throw new ClientException(error.message, error.uri),
+              test: (error) => error is HttpException),
           headers: headers,
-          isRedirect: response.isRedirect,
-          persistentConnection: response.persistentConnection,
-          reasonPhrase: response.reasonPhrase);
+          context: context);
```

Comment:
I don't think it's correct to send the request's context back down the chain as part of the response.